server.js

```diff
@@ -730,167 +729,131 @@ app.post('/submit-match',  asyncMiddleware( async (req, res, next) => {
+        { $inc: $inc },
+        { returnOriginal: false }  // return modified document 
+    )).value;
+
+    // get latest SPRT result
+    const sprt_result = SPRT(match.network1_wins, match.network1_losses);
+    var pending_match_index = pending_matches.findIndex(m => m._id.equals(match._id));
+
+    // match is found in pending_matches
+    if (pending_match_index >= 0) {
+        if (sprt_result === false) {
+            // remove from pending matches
+            console.log("SPRT: Early fail pop: " + JSON.stringify(pending_matches.splice(pending_match_index, 1)));
+            console.log("SPRT: Early fail post-pop: " + JSON.stringify(pending_matches));
+        } else {
+            var m = pending_matches[pending_match_index];
+
+            // remove the match from the requests array.
+            var index = m.requests.findIndex(e => e.seed === seed_from_mongolong(req.body.random_seed));
+            if (index !== -1) {
+                m.requests.splice(index, 1);
+            }
+
+            // update stats 
+            m.game_count++;
+            if (m.network1 == req.body.winnerhash) {
+                m.network1_wins++;
+            } else {
+                m.network1_losses++;
+            }
+
+            // Check > 1 since we'll run to 400 even on a SPRT pass, but will do it at end.
+            //
+            if (sprt_result === true && pending_matches.length > 1) {
+                console.log("SPRT: Early pass unshift: " + JSON.stringify(m));
+                console.log("SPRT: Early pass post-unshift: " + JSON.stringify(pending_matches));
+            }
+        }
+    }
+
+    // Lastly, promotion check!!
     // Check if network2 == best_network_hash and if so, check SPRT. If SPRT pass, promote network1 as new best-network.
     // This is for the case where a match comes in to promote us, after it is no longer the active match in queue.
     //
-    if (!new_best_network_flag && req.body.loserhash == best_network_hash) {
-        db.collection("matches").findOne({ network1: req.body.winnerhash, network2: best_network_hash, options_hash: req.body.options_hash})
-            .then((match) => {
-            // added !is_test flag
-            //
-            if (match && !match.is_test && ( (SPRT(match.network1_wins, match.network1_losses) === true) || (match.game_count >= 400 && match.network1_wins / match.game_count >= 0.55) ) ) {
-                fs.copyFileSync(__dirname + '/network/' + req.body.winnerhash + '.gz', __dirname + '/network/best-network.gz');
-                console.log("New best network copied from (normal check): " + __dirname + '/network/' + req.body.winnerhash + '.gz');
-            }
-        }).catch( err => {
-            console.log("ERROR: " + req.body.winnerhash + " " + best_network_hash + " " + req.body.options_hash);
-            console.log("ERROR: Couldn't check for new best network: " + err);
-        });
+    var best_network_hash = await get_best_network_hash();
```

Comment:
It looks like it's possible for two concurrent wins to both await the best hash to then both decide to promote. Although in this case, it will just end up copying the file twice, so it's probably fine.

Reply:
This seems okay to ignore. The odds we have two promotes happen vs same best network back to back so quickly that get_best_network_hash can't return fast enough seems... unlikely. :)